[guidelines] Writing tests for the Azure SDK (#7784)
Azure/azure-sdk-for-js: documentation/writing_tests.md

For our Engineering Systems to pick up our tests appropriately, our packages must be configured according to their guidelines. In this section we will go through some of these concepts, and provide links that expand them in detail. We will be covering:

- [Engineering goals](#engineering-goals).
- [CI and nightly test configuration](#ci-and-nightly-test-configuration).
- [Delivering live tests to our users](#delivering-live-tests-to-our-users).

### Engineering goals

Though the tests for the Azure SDK for JavaScript and TypeScript must target live resources, we should make sure they only do so when necessary. For this purpose, we must keep in mind the following guidelines:

- Tests should not be flaky. Tests should pass regardless of who's executing it, when they are running, and how many times they run.
- Tests should create the resources they are testing.

> - While writing tests, use your own personal resources for setting up the context in which each test will create their own resources. For example, it is valid to have a static KeyVault while writing KeyVault tests, and then a given test can create a KeyVault Key before validating any of the functionalities of the KeyVault Key. Ask your team to see if there's a resource already in place for test development.

Reviewer: Do we have any recommended tools to help developers avoid committing secrets?

Author: What kind of tool do you mean? We have plenty of tools related to credentials for tests, but I want to understand what you're looking for.

Reviewer: I was wondering if there was some tool that can alert the developer that a password or a service key is about to be committed? 

However, since we're recommending the use of environment variables everywhere this is probably not very important

Reviewer 2: I'm guessing it's going to be `dotenv`, which I think we standardized on all the way into our samples.

Reviewer 3: GitHub is supposed to have some support for cred scanning and I think we can add a CI gate for it? It wouldn't help you not expose the cred (since you already pushed) but it would at least tell you to rotate it.

Reviewer 3: but yes realistically I like always using env vars for this

- Click on "Variable Groups".
- If the variable group you want to select is not visible in that page, add it: You'll see a button that will say "🔗 Link variable group". Click it, then use the "🔍 search" input and type the name `Secrets for Resource Provisioner`, and then link it.

It should end up looking something like this:

![image](https://user-images.githubusercontent.com/417016/72285413-f8e54700-363a-11ea-959e-cb1bc4c074ba.png)

Once the CI is properly configured, you can test that the live tests pipelines work by submitting a comment to a pull request with the name of the pipeline, which will be similar to `js - event-hubs - tests` or `js - keyvault-keys - tests`.

### Delivering live tests to our users

The `test-resources.json` can be used by our users to set up their own test resources. We go through how it's being used with our PowerShell scripts in our [README](https://github.com/Azure/azure-sdk-for-js/blob/master/CONTRIBUTING.md#integration-testing-with-live-services), though we recommend using the same ARM template to expose a "Deploy Button" in the `README.md` of your project. The button will look like this one for KeyVault-Keys:

> [![](http://azuredeploy.net/deploybutton.png)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-sdk-for-js%2Fmaster%2Fsdk%2Fkeyvault%2Ftests-resources.json)

Reviewer: 👍

Now, let's see how we're using the mentioned tools.

### Mocha

Mocha is a well known JavaScript test framework that runs in NodeJS and in the browser. If you're new to Mocha, we recommend you to dive into [Mocha's documentation](https://mochajs.org/).

In this section, we'll explore how the Azure SDK for JS is using mocha, what versions we're using, how we're configuring it, and other recommendations on how to use this library.

#### Mocha in our dependencies

Since we're using [Rush](https://rushjs.io/), we're forced to use the same version of our packages in each one of the projects inside of this repository. If you want to add `mocha` as a dev dependency to a new project inside of this repository, first make sure you are in the root folder of that project, then you can run the following command:

```
rush add --dev -p mocha
```

We're using the following Mocha plugins and related dependencies:

- [`@types/mocha`](https://www.npmjs.com/package/@types/mocha) provides the type definitions for mocha.

> - [`mocha-multi`](https://www.npmjs.com/package/mocha-multi) to let Mocha use multiple reporters.

Reviewer 3: we no longer use `mocha-multi` in our tests thanks to @HarshaNalluru making a custom reporter for us

Author: Thanks! I removed this

#### Mocha in our dependencies

If you want to add `mocha` as a dev dependency to a new project, please use [Rush](https://rushjs.io/) since we're most likely using the same version of Mocha in more projects.once you are in the folder of that project, you can run the following command:

```
rush add --dev -p mocha
```

We're using the following Mocha plugins and related dependencies:

- [`@types/mocha`](https://www.npmjs.com/package/@types/mocha) provides the type definitions for mocha.
- [`mocha-junit-reporter`](https://www.npmjs.com/package/mocha-junit-reporter), which produces JUnit-style XML test results.
- [`nyc`](https://www.npmjs.com/package/nyc) is [Istanbul](https://istanbul.js.org/)'s command line interface.
- [`esm`](https://www.npmjs.com/package/esm), a popular ECMAScript module loader.
- [`source-map-support`](https://www.npmjs.com/package/source-map-support), which provides [source map](https://github.com/mozilla/source-map) support for stack traces in node via the [V8 stack trace API](https://v8.dev/docs/stack-trace-api).

A full `rush add` command that includes `mocha` and all of the previous dependencies follows:

```

> rush add --dev -p mocha @types/mocha mocha-junit-reporter nyc esm source-map-support

Reviewer 4: Does rush support the addition of multiple dependencies at once?

Author: Yes, sir!

Author: You're right! I will split these

- [`karma-env-preprocessor`](https://www.npmjs.com/package/karma-env-preprocessor):
  It's a Karma preprocessor which makes environment variables available to our tests.
- [`karma-json-preprocessor`](https://www.npmjs.com/package/karma-json-preprocessor):
  It's a Karma preprocessor for converting JSON files into JS variables.
- [`karma-json-to-file-reporter`](https://www.npmjs.com/package/karma-json-to-file-reporter):
  It's a Karma reporter that save JSON messages from log to file. We use this to create real local files while executing tests in the browser. It's specifically used by [the recorder](#the-recorder).
- [`karma-junit-reporter`](https://www.npmjs.com/package/karma-junit-reporter):
  It's a Karma reporter for the [JUnit XML format](https://llg.cubic.org/docs/junit/).
- [`karma-mocha`](https://www.npmjs.com/package/karma-mocha):
  It's an adapter for the [Mocha](http://mochajs.org/) testing framework.
- [`karma-mocha-reporter`](https://www.npmjs.com/package/karma-mocha-reporter):
  It's a Karma reporter plugin with mocha style logging.
- [`karma-remap-istanbul`](https://www.npmjs.com/package/karma-remap-istanbul):

>   A Karma reporter that enables remapped reports on watch.

Reviewer 4: ```suggestion
  A Karma reporter plugin that remaps the code coverage report to point to the typescript sources using sourcemappings.
```

- [`karma-chrome-launcher`](https://www.npmjs.com/package/karma-chrome-launcher):

>   Launcher for Google Chrome, Google Chrome Canary and Google Chromium.

Reviewer 4: Are you not adding the launchers for other browsers?
`karma-firefox-launcher`, `karma-ie-launcher`, `karma-edge-launcher`

Author: We're not using these anywhere, or are we? I am at least not aware.

Reviewer 4: These are present in all the storage packages, we do manual testing in these browsers right before the release.

Author: I believe we should document this. Have we made instructions for this anywhere?

Reviewer 4: Storage packages have a requirement to support IE which might not be the case for other services.
Along with IE, firefox and edge are tested too before the release(part of the release checklist for storage libraries).

Maybe you can mention the existence of these karma-launchers and let developers decide if they want to add them for their service.

To install them all together with Karma, you can run the following command:

```
rush add --dev -p karma karma-coverage karma-env-preprocessor karma-json-preprocessor karma-json-to-file-reporter karma-junit-reporter karma-mocha karma-mocha-reporter karma-remap-istanbul karma-chrome-launcher
```

You can see these dependencies at work in our [template project's package.json](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/template/template/package.json).

#### Configuring Karma

Karma is used by our browser test scripts in our `package.json` files. We usually only pass one parameter, `--single-run`, which tells Karma to start and capture all configured browsers, run tests and then exit with an exit code of 0 or 1 depending on whether all tests passed or any tests failed.

```json
  "test:browser": "karma start --single-run",
```

> Karma will default to interpret the file `karma.conf.js`, which should be available at the root of the project. You may specify a different `karma.conf.js` for different test groups, if needed:

Reviewer 4: ```suggestion
Karma will default to interpret the file `karma.conf.js`, which should be available at the root of the project. You may specify a different `karma.conf.js` for different test groups if needed:
```

- In some computers, specially Linux boxes without a graphical environment, tests might not run unless you specifically load a headless Chrome, which should look like this:
```js
  browsers: ["ChromeHeadlessNoSandbox"],
  customLaunchers: {
    ChromeHeadlessNoSandbox: {
        base: "ChromeHeadless",
        flags: ["--no-sandbox"]
    }
  }
```
- Most of our tests focus on checking that our code preemptively works on browsers, even before our services support authentication methods compatible with browsers. For this reason, it's important to disable the browser's web security when necessary. This is done by sending the flag `--disable-web-security`. Combined with `ChromeHeadlessNoSandbox`, it should look like this:
```js
  browsers: ["ChromeHeadlessNoSandbox"],
  customLaunchers: {
    ChromeHeadlessNoSandbox: {
        base: "ChromeHeadless",
        flags: ["--no-sandbox", "--disable-web-security"]
    }
  }
```

> - You might want to avoid logging from the browser, since [the recorder](#the-recorder) uses the logging output to create files. You can omit logs using the `browserConsoleLogOptions` property, which you can conditionally apply, as in the following snippet:

Reviewer 4: ```suggestion
- You might want to avoid logging from the browser, since [the recorder](#the-recorder) uses the logging output to create files(if TEST_MODE is "record"). You can omit logs using the `browserConsoleLogOptions` property, which you can conditionally apply, as in the following snippet:
```

```js
  browserConsoleLogOptions: {

>     terminal: !process.env.TEST_MODE

Reviewer 4: There is no real reason someone would want to use this.
This snippet doesn't have to be in this doc at all if not for the following reasoning. 

- This is tightly coupled with `jsonRecordingFilterFunction` function.
- `jsonRecordingFilterFunction` 
   - works only in `record` mode, no-op in other modes
   - captures all the console logs, prints the console logs straight to the terminal if they are unrelated to recordings, the ones related to recordings are saved into files as the browser recordings.
   - to make the above mechanism work and simultaneously not corrupt the logs, we disable the logging with `browserConsoleLogOptions.terminal` since `jsonRecordingFilterFunction` takes care of logging into the terminal.
- And since the filter-function only works in `record` mode, we disable `browserConsoleLogOptions.terminal` only in record mode. 

This is just to explain how things are working here. I'd suggest you make it `terminal: process.env.TEST_MODE !== "record"` and it lines up with the updated description from my previous suggestion.

Author: Thank you!

Even though tests can be executed in groups, through the use of [Pattern matching test titles](#pattern-matching-test-titles), tests should be considered to be executed all at once. What determines what will be tested will be how tests are written, either through [the recorder](#the-recorder) or through [using conditionals](#using-conditionals). The execution of tests will vary in the `package.json` to provide different build environments, different debugging mechanisms and reportings for automated live tests or automated verification tests. To read more about these topics, you can go to the sections [Configuring Mocha](#configuring-mocha), [Configuring Karma](#configuring-karma) and [CI and nightly test configuration](#ci-and-nightly-test-configuration).

All together, the `package.json` will end up with the following scripts:

```json
  "test": "npm run build:test && npm run test:node && npm run test:browser",
  "test-ci": "npm run build:test && npm run test-ci:node && npm run test-ci:browser",

  "test:node": "mocha --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 180000 --full-trace dist-test/index.node.js",

>   "test-ci:node": "nyc mocha -r esm --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 180000 --full-trace \"dist-esm/test/**/*.spec.js\"",

Reviewer 4: - Are we enforcing this for all the packages?
- So far, we agreed upon `test` command being the convenience for the developers by making `clean`, `build:test` part of it. Are we taking that away?

Author: I will add `npm run clean` too! This is a proposal after all, so I am proposing to have all of what I'm putting here, yes. If that helps.

Author: Here's how these two are looking with your feedback:

```
  "test": "npm run clean && npm run build:test && npm run test:node && npm run test:browser",
  "test-ci": "npm run clean && npm run build:test && npm run test-ci:node && npm run test-ci:browser",
```

Reviewer 4: This is still taking away the convenience of not allowing executing only the node tests with "clean" and "build:test".

Author: Thanks! I've moved them to this:

```json
  "test": "npm run test:node && npm run test:browser",
  "test-ci": "npm run test-ci:node && npm run test-ci:browser",

  // Where the individual commands are:
  "test:node": "npm run clean && npm run build:test && mocha --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 180000 --full-trace dist-test/index.node.js",
  "test-ci:node": "npm run clean && npm run build:test && nyc mocha -r esm --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 180000 --full-trace dist-esm/test/**/*.spec.js",
  // Remember to specify different karma.config.js if needed
  "test:browser": "npm run clean && npm run build:test && karma start --single-run",
  "test-ci:browser": "npm run clean && npm run build:test && npm run test:browser",

  // Where we're cleaning with:
  "clean": "rimraf dist dist-* types",
```

Reviewer 4: Resolving this in favor of the other thread.

Let's see some good and bad examples of reusable code in the three categories described above ([Preparing all of the test cases](#preparing-all-of-the-test-cases), [Preparing some of the test cases](#preparing-some-of-the-test-cases), [Universal utilities](#universal-utilities)).

### Preparing all of the test cases

Re-usable code that will conform a valuable setup for all of the test cases in a given file should be placed in any of the Mocha methods `before` or `beforeEach`. Re-usable code that tears down resources that have been used by all of the test cases should be placed in either `after` or `afterEach`.

Code in either of these methods can be abstracted away into utility functions that may live outside of each test file **only if** these pieces of code are relevant to more than one test file.

One specific example of code that **must** live in the `beforeEach` section is [the recorder](#the-recorder)'s initialization, and later in the `afterEach` section, the recorder's destruction. The following example can be seen in [the recorder's readme](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/test-utils/recorder/README.md):

```ts
  beforeEach(async function() {

>     recorder = record(this);

Reviewer 4: ```suggestion
    recorder = record(this, envSetup);
```

    const credential = await new ClientSecretCredential(
      env.AZURE_TENANT_ID,
      env.AZURE_CLIENT_ID,
      env.AZURE_CLIENT_SECRET
    );

    const keyVaultUrl = "https://myKeyVault.vault.azure.net";
    client = new KeysClient(keyVaultUrl, credential);
  });

  afterEach(function () {
    recorder.stop();
  });
```

### Preparing some of the test cases

Some test cases might need some common setup to properly work. If at least more than one of these tests exist in a single file, their setup can be moved out of each test case into a function (or outside constant), as long as the function (or constant) is declared inside of that same test file, and as long as the common code is sufficiently large. For readability purposes, if the common code is less than four lines long (as an arbitrary reasonable estimate), it will be clearer if it's left repeated in each test case.

Let's define an example in which an SDK client might need several lines of setup in two specific tests:

```ts
describe("some group of functionalities", function() {
  let client;

  beforeEach(function() {

>     recorder = record(this);

Reviewer 4: Same

    client = new Client();
  });

  afterEach(function () {
    recorder.stop();
  });

  it("should test A with a proper setup", async function() {
    client.prepare({
      propertyA: 1,
      propertyB: 2,
      propertyC: 3,
      propertyD: 4,
    });

    const result = await client.A();
    assert.ok(result.value);
  });

  it("should test B with a proper setup", async function() {
    client.prepare({
      propertyA: 1,
      propertyB: 2,
      propertyC: 3,
      propertyD: 4,
    });

    const result = await client.B();
    assert.ok(result.value);
  });
});
```

It is possible to move the common setup code to a function in the same file, as in the following example code:

```ts
describe("some group of functionalities", function() {
  let client: Client;

  beforeEach(function() {

>     recorder = record(this);

Reviewer 4: Same

    const client = new ClientMethodDependency();
    client.A();
    assert.deepEqual(client.history, ["constructor", "A"]);
  });

  it("The ClientMethodDependency's AB method should throw if it's called before A", function() {
    const client = new ClientMethodDependency();
    assert.throws(client.AB);
  });

  it("The ClientMethodDependency's AB method should alter the client's history after A is called", function() {
    const client = new ClientMethodDependency();
    client.A();
    assert.deepEqual(client.history, ["constructor", "A"]);
    client.AB();
    assert.deepEqual(client.history, ["constructor", "A", "AB"]);
  });

});
```

Each test's goal is to verify the behavior of a single functionality, even though some have to rely on a more complex setup than others.

If there are no function calls in a test, a single test case must compare the values of properties with similar names and similar values, for example:

```ts
import { version } from "../some/path/constants.ts";

describe("testing some of the client's public properties", function() {
  let client: Client;

  beforeEach(function() {

>     recorder = record(this);

Reviewer 4: Same

    client = new Client();
  });

  afterEach(function () {
    recorder.stop();
  });

  it("should have a valid version", function() {
    assert.equal(client.version, version)
  });
});
```

It's valid to test more than one property from the result of a single function, like in the following example:

```ts
describe("Tests with more than one property", function() {
  let client: Client;

  beforeEach(function() {

>     recorder = record(this);

Reviewer 4: Same

In addition to being expressive, test titles might have specific hashtags (as in, words preceded by the pound sign), to facilitate pattern matching.

You can see a working example of tests for internal code in our [test/internal/ folder in our template project](../sdk/template/template/test/internal/).

#### Pattern matching test titles

Mocha allows us to only run certain tests as long as they match a given query, with the following arguments:

```
Test Filters
  --fgrep, -f   Only run tests containing this string                   [string]
  --grep, -g    Only run tests matching this string or regexp           [string]
  --invert, -i  Inverts --grep and --fgrep matches                     [boolean]
```

To be able to filter tests through pattern matching with any of these filters, the `describe` and `it` titles can have any hashtag, but preferably one or more of the following:

- `#browser` for tests that are only expected to run in the browser.
- `#node` for tests that are only expected to run on NodeJS.
- `#live` for tests that should only run live and not be recorded, nor played back.

> These and any other hashtags can be used to provide custom `package.json` scripts to run specific sets of tests. We still recommend [using conditionals](#using-conditionals), to ensure that someone reading any test case clearly understands under what conditions this test will be executed or skipped.

Reviewer 4: Other than just about the mocha flags, you might want to add about adding mocha config in karma.conf too to get the same behavior for browser tests
[snippet from karma.conf](https://github.com/HarshaNalluru/azure-sdk-for-js/blob/85a55ab75734e01b1b12f9112eb7611b03a794c1/sdk/eventhub/event-hubs/karma.conf.js#L117-L123)

Author: Thank you!

Author: It should be fine now!

- [Getting feedback](#getting-feedback)

## Engineering setup

The Azure SDK tests not only help us verify that our code is correct, they are also a meaningful way to share how to use our code with our customers. Besides being able to be executed by any developer in the world, they will be triggered by automatic systems that help us verify that we haven't made breaking changes, and also to check that the services we're targeting don't show unexpected behaviors, all of which help us have a better level of confidence before releasing anything to the public.

For our Engineering Systems to pick up our tests appropriately, our packages must be configured according to their guidelines. In this section we will go through some of these concepts, and we will provide links that expand them in detail. We will be covering:

- [Engineering goals](#engineering-goals).
- [CI and nightly test configuration](#ci-and-nightly-test-configuration).
- [Delivering live tests to our users](#delivering-live-tests-to-our-users).

### Engineering goals

Though the tests must target live resources, we should make sure they only do so when necessary. We must keep in mind the following considerations:

> - **Tests should not be flaky.** Tests should pass regardless of who's executing them, when they are running, and how many times they run.

Reviewer 5: Replace "should" with "must" and the guidance for flaky tests is that those tests be skipped and an issue opened up to track fixing the underlying issue. Also, as you alluded to earlier, one should avoid increasing delays in live tests and should instead use polling. 

https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/51/Testing-Guidelines?anchor=best-practices
> Find and eliminate flaky tests. A single flaky test will cause an entire pipeline to fail which means we may have to re-run the pipeline and incur that expense. If you cannot fix the test right now, disable the test and open a work item to fix it later.

> - **Tests should create the resources they are testing**, either through the code of the tests, or through their configuration files.

Reviewer 5: Another way to say this is that "Live testing resources should be created and deleted by automation" 

A link to the deployment scripts' README in each repo might be helpful: https://github.com/Azure/azure-sdk-for-js/blob/master/eng/common/TestResources/README.md

> - **The resources created for the tests should be unique.** Running the same test in parallel, multiple times, should not break them.

Reviewer 5: Similar to this point in our Best Practices: 

https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/51/Testing-Guidelines?anchor=best-practices

> Each live test operates on its own items so that live tests running at the same time do not interfere with each other (use random resource names, etc.)

Author: I do mention this file at the end of this section. Should I move that link up?

Author: I will move this link up.

Author: I moved the link right below the list! Please check this again.

- **Avoid calling to timed delays** (like `setTimeout`) to assert that a change happened in the live resources.

> 

Reviewer 5: Other points from our [Best Practices](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/51/Testing-Guidelines?anchor=best-practices) to consider inserting: 

> Once a live test completes (succeeding or failing) it should clean up any test items it created (Remove-TestResources.ps1 can handle this in scenarios where it is used)

> Live tests should skip execution if there is not sufficient data to execute the test (e.g. credentials to a cloud resource). Put another way: responsibility for whether or not a test should run lies with the test code and not with the larger execution environment (e.g. dotnet test command or the DevOps pipeline running the tests)

Keep in mind that it is valid to have manually created resources during the _development_ of the tests. For example, if you're working with any of the KeyVault clients, you may target a KeyVault created in your personal Azure account. Each one of the individual clients can then assume that a KeyVault provided by environment variables exists, and only focus on creating and managing the resources they're scoped to work with: Keys, Secrets and Certificates. **Ask your team to see if there's a resource already in place for test development.**

> Any resource that is not created by the tests must be defined in an [ARM template](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/overview), so that anyone can build a copy of them. This ARM template will be used by the CI pipelines during builds. Under the context of the previous KeyVault example, to automate the creation of the KeyVault that is passed into the tests, it needs to have been defined specifically in the [CI and nightly test configuration](#ci-and-nightly-test-configuration) files. 

Reviewer 5: > This ARM template will be used by the CI pipelines during builds. 

"This ARM template will be used by the test pipelines for live test runs. "

Live test runs happen nightly or when someone manually queues them.

Author: Thank you!

Regarding delays and `setTimeout`s, please refer to the [using delays](#using-delays) section.

You'll be able to find a more detailed set of recommendations in the section: [Best Practices for writing tests that target live resources](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/51/Testing-Guidelines) [🔒][TIPS].

### CI and nightly test configuration

> To ensure that our tests are executed in the test [Azure DevOps pipelines](https://azure.microsoft.com/en-us/services/devops/pipelines/) that have been previously configured by our team, some configuration files are necessary.

Reviewer 5: > previously configured by our team

These pipelines are mostly generated by automation (except for JS live tests pipelines which are manually created because there are good reasons to use per-package pipelines)

Author: I'm not sure how to express this without making things more confusing!

- [sdk/storage/ci.yml](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/storage/ci.yml).
- [sdk/servicebus/ci.yml](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/servicebus/ci.yml).

For live tests (which an be triggered manually and automatically run every night), we need to provide two files, `test-resources.json` and `tests.yml`.

`test-resources.json` must exist either at the package folder level, or at the service level. This file must contain an [ARM template](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/overview) of the resources needed to run all of the tests of the relevant clients. All of the `test-resources.json` files inside of the service folder will be used in conjunction to deploy resources on every build.

Some examples can be found at: 

- [sdk/keyvault/test-resources.json](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/test-resources.json).
- [sdk/storage/test-resources.json](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/storage/test-resources.json).
- [sdk/servicebus/test-resources.json](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/servicebus/test-resources.json).

> The `tests.yml` file must be placed at the package folder. This file is in charge of specifying when to run the tests for this package, what environments to use to run the tests, and how the tests are executed. You can learn how to write these files by following the guide: [Creating live tests](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/48/Create-a-new-Live-Test-pipeline?anchor=creating-live-tests) [🔒][TIPS].

Reviewer 5: > This file is in charge of specifying when to run the tests for this package 

The nightly schedule is set by automation (or hand) by the engineering systems team. Putting schedules in the yaml file is generally a bad idea because it makes it harder for us to spread out live test runs evenly throughout the day.

Author: Is it better this way? `This file is in charge of specifying what environments to use to run the tests, and how the tests are executed`. I'll push that, but let me know if we should do something different.

First you must go to https://dev.azure.com/azure-sdk/ [🔒][TIPS] and look for the builds that have been configured to target your project, then:

- Click the pipeline you want to test on PR.
- Press the "Edit" button at the top right corner.
- Press the three dot menu at the top right corner. A menu will drop down. Press the "Triggers" option from that menu.
- A page will load with an horizontal menu near the top-center with the following options: "YAML", "Variables", "Triggers" and "History".
- Click the Variables option of that menu.
- Click on "Variable Groups".
- If the variable group you want to select is not visible in that page, add it: You'll see a button that will say "🔗 Link variable group". Click it, then use the "🔍 search" input and type the name `Secrets for Resource Provisioner`, and then link it.

It should end up looking something like this:

> ![image](https://user-images.githubusercontent.com/417016/72285413-f8e54700-363a-11ea-959e-cb1bc4c074ba.png)

Reviewer 5: This is a screenshot of a Unified Pipeline, it should probably be a live tests pipeline.

Author: I'm going to link to [this doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/2/Azure-DevOps-Pipeline-Guidance?anchor=variable-groups) instead of having these steps here.

> Once the CI is properly configured, you can test that the live tests pipelines work by submitting a comment to a pull request with the name of the pipeline, which will be similar to `js - event-hubs - tests` or `js - keyvault-keys - tests`.

Reviewer 5: the comment trigger is `/azp run js - event-hubs - tests`

Author: Thank you!

# Writing tests for the Azure SDK for JS/TS

> The Azure SDK for JavaScript and TypeScript allows users to communicate and control their Azure resources. The development of the Azure SDK should be taken with uttermost care, not only to provide the best API clients to our customers, but also to ensure that the software is reliable through stable, succinct and comprehensible tests. For that purpose, we've made this document that defines how tests should be written. 

Author: I'm not sure why it's not letting me commit these directly, but I'll make sure to take this feedback in! Thank you!

For our Engineering Systems to pick up our tests appropriately, our packages must be configured according to their guidelines. In this section we will go through some of these concepts, and we will provide links that expand them in detail. We will be covering:

- [Engineering goals](#engineering-goals).
- [CI and nightly test configuration](#ci-and-nightly-test-configuration).
- [Delivering live tests to our users](#delivering-live-tests-to-our-users).

### Engineering goals

Though the tests must target live resources, we should make sure they only do so when necessary. We must keep in mind the following considerations:

- **Tests should not be flaky.** Tests should pass regardless of who's executing them, when they are running, and how many times they run.

> - **Tests should create the resources they are testing**, either through the code of the tests, or through their configuration files.

Author: This changed significantly! I'm sorry.

First you must go to https://dev.azure.com/azure-sdk/ and look for the builds that have been configured to target your project, then:

- Click the pipeline you want to test on PR.
- Press the "Edit" button at the top right corner.
- Press the three dot menu at the top right corner. A menu will drop down. Press the "Triggers" option from that menu.
- A page will load with an horizontal menu near the top-center with the following options: "YAML", "Variables", "Triggers" and "History".
- Click the Variables option of that menu.
- Click on "Variable Groups".
- If the variable group you want to select is not visible in that page, add it: You'll see a button that will say "🔗 Link variable group". Click it, then use the "🔍 search" input and type the name `Secrets for Resource Provisioner`, and then link it.

It should end up looking something like this:

![image](https://user-images.githubusercontent.com/417016/72285413-f8e54700-363a-11ea-959e-cb1bc4c074ba.png)

> Once the CI is properly configured, you can test that the live tests pipelines work by submitting a comment to a pull request with the name of the pipeline, which will be similar to `js - event-hubs - tests` or `js - keyvault-keys - tests`.

Author: Thank you!

```

It works by using the Azure endpoint `https://portal.azure.com/#create/Microsoft.Template/uri/` that allows receiving an encoded URL at the end of it. In the KeyVault Keys example, the encoded URL will contain the path of the `test-resources.json`, this one: `https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-sdk-for-js%2Fmaster%2Fsdk%2Fkeyvault%2Fkeyvault-keys%2Ftest-resources.json`.

Once clicked, the deploy button will load a form at Azure that should ask some basic information, and then allow anyone to deploy the same set of resources, already properly configured, to their accounts. This form is automatically generated from the ARM template, so to help our users go through it in detail, and also to inform them of the resources they will be creating, we recommend writing these details in a new file in your project's folder, called `TEST_RESOURCES_README.md` and linking it from your `README.md`. Here's an example that applies to all of our KeyVault clients: https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/TEST_RESOURCES_README.md

## Recommended tools

> Writing tests for JavaScript and TypeScript requires testing tools, such as a test framework, an assertion library, and a way to bundle and run these tests in various environments. The JavaScript community has many overlapping tools that one could pick to fulfill any of these tasks. To ensure that the testing experience across the Azure SDK for JS/TS is consistent and reliable, we've picked the following external testing tools:

Author: Thank you!

> #### Tips for reading this document
> - If you see any link that ends in a lock [🔒][TIPS], that link is private, only accessible to developers members of the Azure SDK team.
> - Links that end up with a clock [⏲][TIPS], these are links that might contain outdated information. The information that's being referred might move elsewhere as time passes. Make sure to find the latest version of the linked file, or the content of the file that is relevant to what's being described in this document.

[TIPS]: #tips-for-reading-this-document

## Index

- [Engineering setup](#engineering-setup)
    - [Engineering goals](#engineering-goals)
    - [CI and nightly test configuration](#ci-and-nightly-test-configuration)
    - [Delivering live tests to our users](#delivering-live-tests-to-our-users)
- [Recommended tools](#recommended-tools)
    - [Mocha](#mocha)
        - [Mocha in our dependencies](#mocha-in-our-dependencies)
        - [Configuring Mocha](#configuring-mocha)
        - [Code coverage with Mocha and nyc](#code-coverage-with-mocha-and-nyc)
        - [Handling timeouts](#handling-timeouts)

>         - [On the usage of before, beforeEach, after and afterEach](#on-the-usage-of-before--beforeeach,-after-and-aftereach)

Author: Fixing it!

Mocha's `before` and `beforeEach` are methods that allow you to specify functions that should be executed either before all of the tests run or before each test runs, respectively. Similarly, `after` and `afterEach` exist so that developers can specify functions that should be executed after all tests run or after each test runs.

We recommend using `beforeEach` rather than `before`, just as much as we recommend using `afterEach` rather than `after`. The idea is that each test case should not depend on a state that is shared with other tests. Use `beforeEach` to execute tasks that will prepare the resources needed for each test to run cleanly, and `afterEach` to tear down or clean those settings before the next test runs.

`before` and `after` can be used to define or create heavy resources that could cleanly be used by more than one test. We will see some examples below.

This discouraged example of `before` and `after` shows that neither clients nor stateful objects should be assigned in these functions ([source](https://github.com/Azure/azure-sdk-for-js/blob/b3bc4375c038605e00797fd45b1efb9bff4ea651/sdk/template/template/test/internal/beforeAfter.spec.ts#L14) [⏲][TIPS]):

```ts

> describe("Discouraged example of `before` and `after`", function() {

Author: I'm up for a better idea!

    // Do not do this. Use async/await
    myAsyncMethod.then(done);
  });

  it.skip("Discouraged async test with arrow async function", async () => {
    // Do not use arrow functions to declare test cases,
    // tests will lose Mocha's context, which is useful, and needed by the Recorder.
    await myAsyncMethod();
  });

  it("Recommended async test with natural async function", async function() {
    // This is good ✅
    await myAsyncMethod();
  });
});
```

### Chai

Chai is an assertion library, similar to Node's built-in `assert`. While `assert` is a module that belongs to NodeJS's standard library, it was mainly designed to write tests for the NodeJS's core, and [it has been discouraged from being used as a dependency](https://github.com/nodejs/node/issues/4532). Chai is very popular, and it is widely used for both NodeJS and the browser. You can read more about Chai through their main website: https://www.chaijs.com/

Chai makes testing much easier by providing an extensive list of assertions that can run against your code. This list of assertions can be seen in detail by going to their assertion style guide: <https://www.chaijs.com/guide/styles/>

> Another important learning resource for Chai is: Chai Assertions for Promises <https://www.chaijs.com/plugins/chai-as-promised/>.

Author: Nope! Removing this.

#### Recommended Chai assertion style

The Azure SDK for JavaScript (and TypeScript) prefers Chai's Assert style. For example:

```ts
import { assert } from "chai";

let foo = "bar";
let beverages = { tea: [ 'chai', 'matcha', 'oolong' ] };

assert.typeOf(foo, "string");
assert.typeOf(foo, "string", "foo is a string");
assert.equal(foo, "bar", "foo equals to `bar`");
assert.lengthOf(foo, 3, 'foo`s value has a length of 3');
assert.lengthOf(beverages.tea, 3, 'beverages has 3 types of tea');
```

#### Chai in our dependencies

Since we're using [Rush](https://rushjs.io/), we have to use the same version of our packages in each one of the projects inside of this repository. If you want to add `chai` as a dev dependency to a new project inside of this repository, first make sure you are in the root folder of that project, then you can run the following command:

```
rush add --dev -p chai
```

Since we're using TypeScript, make sure to also add the types:

```
rush add --dev -p @types/chai
```

> ### Rollup

Author: When you say to separate this, do you mean in several documents? I can make a `test_guidelines` folder and split this one in several files, and link them through the `index.md`. Would that help?

> ### Rollup

Reviewer 3: I think *maybe* it would help, since this document is pretty big and that can spook people. 

On the other hand, it's kinda nice that it's one document and searchable, so I don't feel strongly about this.

> ### Rollup

Author: I'll leave this open for debate!

Rollup is a module bundler for JavaScript. It uses the new standardized format for code modules included in the ES6 revision of JavaScript. This will eventually be possible natively everywhere, but Rollup lets you do it today.

We use Rollup to take the result of the TypeScript compiler and carefully bundle it differently depending on the target platform (NodeJS or the browsers) with:

- The source maps available from our compiled TypeScript and our dependencies.
- Without specific sections of code that exist only for a specific platform.
- To transform the necessary dependencies from our `node_modules`, from CommonJS to ES6. 
- With a banner including Microsoft's copyright at the top of the generated file.
- With the consideration necessary to output a working JavaScript file compatible with browsers.
- With special settings necessary to run our tests in the browsers.

If you want to learn more about Rollup, you can read the Rollup guide at: https://rollupjs.org/guide/en/

#### Rollup in our dependencies

> Since we're using [Rush](https://rushjs.io/), we're forced to use the same version of our packages in each one of the projects inside of this repository. If you want to add `chai` as a dev dependency to a new project, once you're at the root of this new project, you can run the following command:

Author: Changing this to:

> If you want to add `chai` as a dev dependency to a new project, please use [Rush](https://rushjs.io/) since we're most likely using the same version of Chai in more projects. Once you're at the root of this new project, you can run the following command:

### The Recorder

The Azure SDK for JavaScript and TypeScript uses a custom utility to record tests that hit the live endpoints, so that these tests can be executed almost instantly against these recordings instead of hitting the live services. This tool is called [@azure/test-utils-recorder](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/test-utils/recorder). It's an unpublished package that can be added using rush, as follows:

```
rush add --dev -p @azure/test-utils-recorder
```

You can read more about the recorder in its readme: https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/test-utils/recorder

## Test folder structure

Tests for the Azure SDK for JavaScript and TypeScript should be all executed using the NPM command `test`, they should be stored in the `test` folder, they must be written in TypeScript, in files that will end in `.spec.ts`. Our [recommended tools](#recommended-tools) should provide the test framework and proper ways of compiling and bundling our tests correctly. They should be able to run and be debugged on the environments we're targeting, which are currently only NodeJS and the browsers.

> Even though tests can be executed in groups, through the use of [Pattern matching test titles](#pattern-matching-test-titles), tests should be considered to be executed all at once. What determines what will be tested will be how tests are written, either through [The Recorder](#the-recorder) or through [using conditionals](#using-conditionals). The execution of tests will vary in the `package.json` to provide different build environments, different debugging mechanisms and reportings for automated live tests or automated verification tests. To read more about these topics, you can go to the sections [Configuring Mocha](#configuring-mocha), [Configuring Karma](#configuring-karma) and [CI and nightly test configuration](#ci-and-nightly-test-configuration).

Author: Thanks!

> Even though tests can be executed in groups, through the use of [Pattern matching test titles](#pattern-matching-test-titles), tests should be considered to be executed all at once. What determines what will be tested will be how tests are written, either through [The Recorder](#the-recorder) or through [using conditionals](#using-conditionals). The execution of tests will vary in the `package.json` to provide different build environments, different debugging mechanisms and reportings for automated live tests or automated verification tests. To read more about these topics, you can go to the sections [Configuring Mocha](#configuring-mocha), [Configuring Karma](#configuring-karma) and [CI and nightly test configuration](#ci-and-nightly-test-configuration).

Author: Moved everything to `the recorder` except in the index and in the headline of the recorder's section.

All together, the `package.json` will end up with the following scripts:

> ```json

Author: I've separated this section in "Package configuration" and "Test folder structure". Let me know if that helps!

  // And integration-test means...
  "integration-test:browser": "echo skipped",
  "integration-test:node": "nyc mocha -r esm --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 180000 --full-trace \"dist-esm/test/**/*.spec.js\"",
```

Besides the target environment, tests may or may not target live resources, and either should target public or internal code. To navigate through these concepts, we will make clear distinctions of what we will be testing, based on three main aspects:

- [Testing cloud resources](#testing-cloud-resources).
- [Public or internal tests](#public-or-internal-tests).
- [Testing API functionalities](#testing-api-functionalities).

We will continue by elaborating on these three distinctions in the next section.

Keep in mind that [test titles](#test-titles) may include information about the environment they're targeting, what resources they're testing and any other distinction up to the good judgement of the developer. For more information, check out the section: [Writing test cases](#writing-test-cases).

### Testing cloud resources

> By making use of [the recorder](#the-recorder), we're allowed to write tests targeting real live resources, and then run them instantly against a local copy of the server responses by specifying the `TEST_MODE` environment variable to either `record`, `live` or `playback`. Not specifying the environment variable will be considered `playback`, since we want to let anyone see the tests running by default without first having to set up real resources.

Author: I'll use `the recorder`.

### Public or internal tests

Tests can also either target internal code or code that is part of the public API of our projects. We will separate where these tests are located by dividing them in folders, as follows:

> ```

Author: Define much earlier!

> ```

Author: I've separated this section in "Package configuration" and "Test folder structure". Let me know if that helps!

  });

});
```

Each test's goal is to verify the behavior of a single functionality, even though some have to rely on a more complex setup than others.

If there are no function calls in a test, a single test case must compare the values of properties with similar names and similar values, for example:

```ts
import { version } from "../some/path/constants.ts";

describe("testing some of the client's public properties", function() {
  let client: Client;

  beforeEach(function() {
    recorder = record(this);
    client = new Client();
  });

  afterEach(function () {
    recorder.stop();
  });

  it("should have a valid version", function() {
    assert.equal(client.version, version)
  });
});
```

It's valid to test more than one property from the result of a single function, like in the following example:

```ts
describe("Tests with more than one property", function() {
  let client: Client;

  beforeEach(function() {
    recorder = record(this);
    client = new Client();
  });

  afterEach(function () {
    recorder.stop();
  });

  it("should test A", async function() {
    const result = await client.A();
    assert.ok(result.value);

>     assert.ok(result.createdAt instanceof Date);

Reviewer 3: nit:  should use `assert.typeOf` or `assert.instanceOf`

https://www.chaijs.com/api/assert/#method_typeof 
https://www.chaijs.com/api/assert/#method_instanceof

```ts
describe("Tests for the internal classes", function() {

  describe("Tests for the InternalClass", function() {
    it("The InternalClass should be able to be initialized", function() {
      const Internal = new InternalClass();
      assert.exists(Internal);
    });

    it("The InternalClass's returnsTrue should return true", function() {
      const Internal = new InternalClass();
      const result: boolean = Internal.returnsTrue();
      assert.isTrue(result);
    });
  });

  describe("Tests for the InternalInheritedClass", function() {
    it("The InternalInheritedClass should be able to be initialized", function() {
      const Internal = new InternalInheritedClass();
      assert.exists(Internal);
    });

    it("The InternalInheritedClass's returnsTrue should return true", function() {
      const Internal = new InternalInheritedClass();

>       const result: boolean = Internal.returnsTrue();

Reviewer 3: in general do we recommend testing non-shadowing inherited members? seems unnecessary

Author: I'm mostly concerned about displaying the functionality to our users. Is it enough to let them figure out or would it help being verbose about what each class is capable of doing?

Reviewer 3: I guess my instinct is you don't bother testing inherited members unless you do something that uniquely changes the way they act. Otherwise you're just testing that inheritance works, which hopefully isn't something we need to do.

Author: I think what you're saying needs to be explicitly mentioned in the guidelines! I'll make sure it is, and I will make sure the tests reflect it 🌞

Author: I think I addressed this now! I'll resolve this. Let me know if I should change it!

If a method can't use `core-lro` (for example, if the service cannot be trusted to provide signs of completion, or finish within reasonable time), and thus we see the need to use a different delay strategy, we should always wait until the next possible operation can be fulfilled. **Tests that need to wait for an operation that cannot be trusted to finish reasonably should not run during the automated execution of live tests**. The following snippet can be used to skip these tests in live mode:

```ts
  // `isRecordMode` and `isPlaybackMode` are methods exported from the Recorder.
  if (!isRecordMode() && !isPlaybackMode()) {
    return this.skip();
  }
```

For example, in `@azure/keyvault-keys` we provide a method to purge keys that has not been moved to use `core-lro`. To check that it has finished, we do a while loop where we try to make the next operation until it passes:

```ts
import { delay, isRecordMode, isPlaybackMode } from "@azure/test-utils-recorder";
// ...

describe("Keys client - restore keys and recover backups", () => {
  // ...
  it("can restore a key with a given backup", async function() {

    // This test can't be expected to finish in any reasonable time.

>     if (!isRecordMode() && !isPlaybackMode()) {

Reviewer 3: is there no `isLiveMode()`?

Author: There actually is! Thanks for reminding me.

      return this.skip();
    }

    // ...
    await client.createKey(keyName, "RSA");
    const backup = await client.backupKey(keyName);
    await client.purgeDeletedKey(keyName);
    while (true) {
      try {
        await client.restoreKeyBackup(backup as Uint8Array);
        break;
      } catch (e) {
        console.log("Can't restore the key since it's not fully deleted:", e.message);
        console.log("Retrying in one second...");
        // This delay method comes from the recorder

>         await delay(1000);

Reviewer 3: seems like this delay number should be its own const

Author: Interesting! Why?

Reviewer 3: oh I think this is just me having the general sense that random number literals in code are to be avoided because it's not clear what they mean or if they get used in multiple places it's hard to update.

e.g.

```typescript
const KEY_DELETION_DELAY_IN_MS = 1000;
```

but I don't feel too strongly about this since it's the only usage of the literal.

Author: Pushing this!

      }
    }
    const getResult = await client.getKey(keyName);
    assert.equal(getResult.name, keyName, "Unexpected key name in result from getKey().");
    // ...
  });
  // ...
});
```

Keep in mind that Mocha will have a timeout configuration that will prevent this to run up forever.

The specific `delay` method used in the code above comes from [the recorder](#the-recorder), so that in playback, there will be no delay at all, and tests will pass as soon as possible.

### Exceptions and edge cases

While testing the Azure SDK clients for JavaScript and TypeScript, we should document all known exceptions through the use of `@throws` in the TypeDoc documentation. Tests for exceptions should only exist if these exceptions have been implemented within the code of the SDK clients, and not if they happen on the service side.

> Similarly, since the public API surface of our clients contain a large set of properties resulting from any of the methods that our clients implement, our tests should not focus on verifying that all of the client properties exist or are expectedly received. Tests should include only as many properties as it can be relevant for the use case that each test case is representing. For this purpose we should also take advantage of strict types. If our types can be descriptive and thorough (, and our internal code is not skipping any types by using `any`), we will be able to trust that our API is behaving reasonably well.

Reviewer 3: ```suggestion
Similarly, since the public API surface of our clients contain a large set of properties resulting from any of the methods that our clients implement, our tests should not focus on verifying that all of the client properties exist or are expectedly received. Tests should include only as many properties as it can be relevant for the use case that each test case is representing. For this purpose we should also take advantage of strict types. If our types can be descriptive and thorough (and our internal code is not skipping any types by using `any`) we will be able to trust that our API is behaving reasonably well.
```

> To use `tsdoc` to document our expected API exceptions, use `@throws`. The implementation can be seen ]here (includes examples)](https://github.com/microsoft/tsdoc/pull/175).

Reviewer 3: broken link

Author: Thanks!

  It's a Karma preprocessor for converting JSON files into JS variables.
- [`karma-json-to-file-reporter`](https://www.npmjs.com/package/karma-json-to-file-reporter):
  It's a Karma reporter that save JSON messages from log to file. We use this to create real local files while executing tests in the browser. It's specifically used by [the recorder](#the-recorder).
- [`karma-junit-reporter`](https://www.npmjs.com/package/karma-junit-reporter):
  It's a Karma reporter for the [JUnit XML format](https://llg.cubic.org/docs/junit/).
- [`karma-mocha`](https://www.npmjs.com/package/karma-mocha):
  It's an adapter for the [Mocha](http://mochajs.org/) testing framework.
- [`karma-mocha-reporter`](https://www.npmjs.com/package/karma-mocha-reporter):
  It's a Karma reporter plugin with mocha style logging.
- [`karma-remap-istanbul`](https://www.npmjs.com/package/karma-remap-istanbul):
  A Karma reporter plugin that remaps the code coverage report to point to the typescript sources using sourcemappings.
- [`karma-chrome-launcher`](https://www.npmjs.com/package/karma-chrome-launcher):
  Launcher for Google Chrome, Google Chrome Canary and Google Chromium.

You can install them with:

```

> rush add --dev -p karma

Reviewer 4: Does this make sure that the version being installed is in sync with the rest of the SDKs in the repo?

Author: Rush tries to do this by default:

> If no version is specified, a version will be automatically detected (typically
either the latest version or a version that won't break the
"ensureConsistentVersions" policy).

From: https://rushjs.io/pages/commands/rush_add/

Let's define an example in which a SDK client might need several lines of setup in two specific tests:

```ts
describe("some group of functionalities", function() {
  let client;

  beforeEach(function() {
    client = new Client();
  });

  it("should test A with a proper setup", function() {
    client.prepare({
      propertyA: 1,
      propertyB: 2,
      propertyC: 3,
      propertyD: 4,
    });

    const result = await client.A();
    assert.ok(result.value);
  });

  it("should test B with a proper setup", function() {
    client.prepare({
      propertyA: 1,
      propertyB: 2,
      propertyC: 3,
      propertyD: 4,
    });

    const result = await client.B();
    assert.ok(result.value);
  });
});
```

It is possible to move the common setup code to a function in the same file, as in the following example code:

```ts
describe("some group of functionalities", function() {
  let client: Client;

  beforeEach(function() {
    client = new Client();
  });

  function defaultPrepareClient(client: Client) {
    client.prepare({
      propertyA: 1,
      propertyB: 2,
      propertyC: 3,
      propertyD: 4,
    });
  }

>   it("should test A with a proper setup", function() {

Reviewer 2: Needs to be async? Same for the test below.

Author: I've addressed all of the async functions now! I think.

    defaultPrepareClient(client);
    const result = await client.A();
    assert.ok(result.value);
  });

  it("should test B with a proper setup", function() {
    defaultPrepareClient(client);
    const result = await client.B();
    assert.ok(result.value);
  });
});
```

### Universal utilities

In case that a specific utility function might be useful for more than one test case, and more than one test file, and even potentially in more than one project, this code should be moved away from each client into a common private package in the [test-utils](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/test-utils/) folder. If the package doesn't exist, it should be created.

Let's say we have a test utility that we call `retry.ts`, with the following content:

```ts

> import { delay as coreDelay } from "@azure/core-http";

Reviewer 2: Note that in another librarywe're using 'delay' from test-utils. Is there any reason to use one vs the other when it comes to writing test code?

Author: I will make this use the recorder's delay. Thanks!

Author: Done!

/**
 * A simple abstraction to retry, and exponentially de-escalate retrying, a
 * given async function until it is fulfilled.
 * @param {() => Promise<T>} target The async function you want to retry
 * @param {number} delay The delay between each retry, defaults to 1000
 * @param {number} timeout Maximum time we'll let this lapse before we quit retrying, defaults to Infinity
 * @param {number} increaseFactor Increase factor of each retry, defaults to 1
 * @returns {Promise<any>} Resolved promise
 */
export async function retry<T>(
  target: () => Promise<T>,

>   delay: number = 1000,

Reviewer 2: Maybe just name this `delayInMs` (like we force in our API anyways) and then you don't have to rename `delay` to `coreDelay`

Author: I will take your advice here!

Author: Done!

### Inner parts of a test

While writing test cases, it will be easier to think of the test case's contents as if there were three clear divisions:

- The first lines of the test should deal with setting up the necessary variables and values that will be used for the main functionality that will be tested.
- The body of the test should focus on using these values to trigger the main functionality that is being tested.
- The last set of lines should be focused on comparing the results obtained, to confirm that they are in the expected shape, or contain the expected values.

### Using conditionals

Ideally, each test case should only have one possible behavior. To minimize unexpected outcomes, conditionals should be avoided. The only case in which conditionals are encouraged is to specify when to skip a test. Some tests might only run in a browser, or only run in NodeJS, or only run live (no record or playback), for this purpose, conditionals can be used. For example:

```ts
import { isNode } from "@azure/core-http";

describe("testing the client's basic methods", function() {
  let client: Client;

  beforeEach(function() {

>     client = new Client();

Reviewer 2: No cleanup of client needed? If it's stateless do we need a new one each time?

Author: I will move it to "before".

Author: So, it depends of where the recorder is defined. Right now, the recorder must be defined in `beforeEach`, since the authentications can mess up playbacks. I've added more context on these beforeEach, as in: a snippet of the definition of the recorder and an afterEach stopping the recorder.

Author: We also don't have many clients that require cleanup. Do we? I'm trying to remember.

  });

  it("should test A #browser", function() {

>     if (isNode) {

Reviewer 2: Ah..why is this needed? The tag isn't enough?

Author: Tags can be used to specify what tests to run, which can be used in the configuration files, but at a minimal case tags won't do anything special. This section is about when to use conditionals!

      return this.skip();
    }
    // Test contents..
  });

  it("should test B #node", function() {
    if (!isNode) {
      return this.skip();
    }
    // Test contents..
  });

>   it("should test C #live", function() {

Reviewer 2: Just trying to understand your system here - are you saying that any test that doesn't say #live will activate recordings and run during the unit-test phase?

Reviewer 2: Also, does this imply that #live tests always run in both the browser and in node? Or do I combine flags?

Author: This is described in more detail a bit before this snippet. I mean here: https://github.com/Azure/azure-sdk-for-js/pull/7784/files/870108e6e0b34e20f65796a0bd379c3d774e1aa0#diff-a43d448b275343b7d964b6e68f7cabf4R840

Hashtags are optional though. One "can" use them.

    if (process.env.TEST_MODE) {
      return this.skip();
    }
    // Test contents..
  });
});
```

In case conditionals might appear to be necessary in other scenarios, consider separating the test case into as many test cases as necessary first.

### Using delays

The API methods that we provide should always provide a way for users to know when an operation finishes. For this reason, any method that doesn't immediately respond with a completed operation should use our [Long Running Operations strategy (core-lro)](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/core/core-lro).

If any method can't use `core-lro` within reasonable time, and thus we are inclined to use a different delay strategy, we should always wait until the next possible operation is can be fulfilled.

> For example, in KeyVault-Keys we provide a method to purge keys that has not been moved to use `core-lro`. To check that it has finished, we do a while loop where we try to make the next operation until it passes:

Reviewer 2: ```suggestion
For example, in `@azure/keyvault-keys` we provide a method to purge keys that has not been moved to use `core-lro`. To check that it has finished, we do a while loop where we try to make the next operation until it passes:
```

```ts
import { delay } from "@azure/test-utils-recorder";
// ...

describe("Keys client - restore keys and recover backups", () => {
  // ...
  it("can restore a key with a given backup", async function() {
    // ...
    await client.createKey(keyName, "RSA");
    const backup = await client.backupKey(keyName);
    await client.purgeDeletedKey(keyName);

>     while (true) {

Reviewer 2: So in EH, where tests could potentially take an indeterminate amount of time (by accident) we have a loopUntil() method that basically retries a set # of times with a delay so that way you don't run eternally.

Would you suggest something along that nature here? (at the very least, a counter?)

Author: That's a great point and that's why I submitted the while, to get smarter responses. We did have a retry tool, but these tools need to be moved to a common test utilities package, so that other packages can use it.

We do use whiles for now since our retry utility was bugged and would get the test stuck forever. In theory, mocha's timeout should handle the exceptional case in which we might not get resolved on an appropriate time, though I do agree that a common utility will help. We should think together on a common set of test utilities. That work is perhaps separate from the initial test guidelines approach (or not! Up for debate).

Author: After some thought, we should not have things like my retry tool, or loopUntil, at all. These tests can exist as temporary measures and not run on live test runs. All of our methods that don't return instantly should use `core-lro`. If we can't make them use core-lro because of a service limitation, we should report this issue to the service team and continue pushing until they provide us an approach that we can reasonably make into a LRO.

      try {

>         await client.restoreKeyBackup(backup as Uint8Array);

Reviewer 2: Unrelated to this testing sample - why is that the result from .backupKey can't be passed, uncasted, to restoreKeyBackup? Are they not normally paired together?

Author: Oh the issue is that backup can be undefined. Would it be better if I use the exclamation sign?

Reviewer 2: Probably my lack of understanding in KeyVault - how does 'backup' ever become 'defined' then (it's const, right?). 

Isn't the point of this to restore a key you've backed up (ie, one that definitely did exist?).

Author: it's const, but the API can either return a backup or undefined. Perhaps we should throw if there's no backup! I'll bring this up to the KeyVault team.

        break;
      } catch (e) {

>         console.log("Can't restore the key since it's not fully deleted:", e.message);

Reviewer 2: For testing purposes - we should check that the error that came back is actually a 409 (or whatever comes back if the key is still "present"), correct?

Author: This while is to test that the deleted key has been purged. There's no current way to know if the key was finally purged. We've been submitting feedback to the service team.

Reviewer 2: Is 'e' an HTTP error? Is the status code useful for this?

Author: nope. There's no way to know when a key is purged at the moment. Our existing LROs use status codes like how you mention, but we can't make a LRO for this one. We've already submitted the feedback to the KeyVault service team.

        console.log("Retrying in one second...");
        // This delay method comes from the recorder
        await delay(1000);
      }
    }
    const getResult = await client.getKey(keyName);
    assert.equal(getResult.name, keyName, "Unexpected key name in result from getKey().");
    // ...
  });
  // ...
});
```

Keep in mind that Mocha will have a timeout configuration that will prevent this to run up forever.

The specific `delay` method used in the code above comes from the [Recorder](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/test-utils/recorder), so that in playback, there will be no delay at all, and tests will pass as soon as possible.

### Exceptions and edge cases

> While testing the Azure SDK clients for JavaScript and TypeScript, we should document each client method exceptions through the use of `@throws` in the TypeDoc documentation, and avoid writing test cases for exceptions. Test cases should focus on demonstrating the functionalities of the client, in relation to the functionalities of the service. Exceptions should be in principle documented.

Reviewer 2: > avoid writing test cases for exceptions

If we document it (ie, it's part of the customer facing interface) we should have a test asserting that it happens, right? (when possible). Is there a reason we shouldn't write test cases like this?

Author: The main issue is that many of the possible service exceptions could be hard to replicate, and we don't want to write mocks representing each one of these cases for now, at least from the feedback I've received. How would we test all of the exceptions? If it's through mocks, do we want to go that way? I'll leave it open to get more feedback, for sure.

Reviewer 2: You've got a good point - I think 'avoid writing test cases for exceptions' should be explained a bit so people can use their discretion. In general though, I agree with you.

The exceptions (minus the message text itself and more any programatically actionable codes) should be tested when we can.

Author: I believe that at a minimum, exceptions MUST be tested if they are representative of use cases written by customers. The phrase "exceptions should be tested when we can" does not provide a clear distinction of when this should happen. If we were developing the services, these edge cases would be tested on the services side. We should demand them to provide clear documentation of when exceptions happen. Some of this might currently be available, and based on that we can document how these APIs behave.

Having said that, I do feel odd to avoid writing tests for all of the exceptions, so please consider my response as a request to narrow down what exceptions to test with more clarity.

Reviewer 3: Maybe it makes sense to exercise exceptions we throw in the convenience layer / in helper functions to make sure that guards/checks are working as expected, but not to exercise exceptions that come from the service (e.g. RestError and friends.)

Author: That seems good to me! I'll make sure the document reflects that! Thank you, @xirzec

> While testing the Azure SDK clients for JavaScript and TypeScript, we should document each client method exceptions through the use of `@throws` in the TypeDoc documentation, and avoid writing test cases for exceptions. Test cases should focus on demonstrating the functionalities of the client, in relation to the functionalities of the service. Exceptions should be in principle documented.

Reviewer 2: ```suggestion
While testing the Azure SDK clients for JavaScript and TypeScript, we should document each client method exceptions through the use of `@throws` in the TypeDoc documentation, and avoid writing test cases for exceptions. Test cases should focus on demonstrating the public API of the service. Exceptions should be documented.
```

(not totally sure about this edit, just saw functionalities twice and thought maybe  it wasn't needed)

Similarly, the public API surface of our clients will contain a large set of properties resulting from any of the methods that our clients implement. Tests should not focus on verifying that each property of of our clients exist. While tests can check that the values of properties are expected, they should include only as many properties as it can be relevant for the use case that each test case is representing. For this purpose we should also take advantage of strict types. If our types can be descriptive and thorough, and our internal code is not skipping through the types (through the use of `any`), we will be able to trust that our API is behaving reasonably well.

As a final note on this regard, as we develop clients for the SDKs, we will encounter edge cases in the services we are testing. Edge cases should be reported and documented, but our tests should preferably avoid having to go through them, unless we can identify some of these cases as valuable resources for our customers.

### How tests should look

> In general, the tests of the Azure SDK for JavaScript and TypeScript should be considered useful resources that demonstrate how to use the functionalities that our clients offer, and how these are expected to behave. Our test cases should assert that we are providing well constructed features to our users, and our users should be able to go through our tests, understand them with minimal effort, and use our test code to their advantage. Our tests should therefore be _empowering everyone_.

Reviewer 2: Example? If you're going to write about how something looks you have to show it :)

Author: I've added a list of questions that can be used to make a self assessment of how good are our tests. Let me know if that helps!

## Getting feedback

> Writing tests the right way can be quite challenging. Make sure to make pull request throughout the process and ask for feedback from your team. Ask them questions about how easy is to follow through, from the perspective of a new user. Once you write them, monitor how they behave through our Engineering Systems. Don't be afraid to ask our Engineering Systems team about feedback to your tests. Making your tests constantly better is your responsibility as a developer, but you have all of your team at your back. Together, we can make our tests are as good as they can possibly be.

Reviewer 2: What kind of feedback are you thinking we'd get from Engineering Systems? I'd not considered talking to them about it.

Is it like the runtime/memory requirements of tests? Duration?

Author: What resources we're creating, how much time the tests take, etc.

Reviewer 2: Makes total sense - I would definitely include those kinds of things in the paragraph.

Author: I will update the paragraph based on your feedback!

Author: I have updated this based on your feedback!

```json
    "integration-test:node": "find dist-esm/test -name '*.spec.js' | xargs nyc mocha -r esm --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 180000 --full-trace",
    "unit-test:node": "mocha --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 180000 --full-trace dist-test/index.node.js",
```

Keep in mind that Mocha will be directly called from our `package.json` scripts only for our **NodeJS** tests. For our browser tests, we will be using [Karma](#karma).

You can look at how Mocha's configuration is present in our [template project's package.json file](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/template/template/package.json).

#### Code coverage with Mocha and nyc

> Our `integration-test` script will output code-coverage on our nightly and release builds. It's recommended to run this manually to confirm that code coverage is high. Code coverage should be above 80%, though code coverage should not change how we write code or tests.

Reviewer 4: > Code coverage should be above 80%

Maybe we should focus on this.. 
Ideally, it should be 100% and we need to ensure that each and every branch is covered.

Reviewer 3: In my experience when you go for 100% you end up writing useless tests that get the coverage but don't test anything useful. Remember coverage doesn't imply that you assert anything, just that code ran. :)

Reviewer 2: In particular this being an integration test means that 100% isn't feasible unless you can provoke every error condition, etc...

Unit-tests, however, should have a higher goal IMO. We just need to be reasonable about not requiring coverage in generated code, accessors, etc...

Author: We don't require code coverage. I've expanded on this topic to reflect that 80% is considered a good measure, but it's up to the good judgement of developers. If anybody has a clear strategy that can be enforced, please let us know.

#### Handling timeouts

> All of the tests in the Azure SDK repository should have a timeout. In the previous section, we have discussed how this timeout is defined for _mocha_, though the parameter `--timeout` with an additional milliseconds amount, which is sent during the invocation of mocha. Now we will define how to know how many milliseconds to pass to this parameter.

Reviewer 4: I think our pipelines themselves have a default timeout too and mocha cannot go beyond them.

Reviewer 2: @jeremymeng was asking about this the other week so he can probably provide context on how you actually change it (or are supposed to change it).

Reviewer 6: The build pipeline timeout is for the total time of the build.  This timeout we passed to mocha is the time out for one test (e.g., one `it()` function). I think what @sadasant has here is proper as it's taking from the slowest passing test from the history.  I don't believe we need to take the sum of running times in the worst case and make sure the sum is below build pipeline timeout limit.

Author: Thank you, Jeremy.

This timeout strategy is taken in coordination with Engineering.

Something to consider is that if we ever reach the engineering timeout (of one hour) all of our logs will most likely disappear. I think this might be an Engineering bug so I'm not sure if I should document this in the guidelines. The sum of our individual timeouts should still be far from reaching the Engineering's timeout. It will be easy to detect if we ever need to increase the Engineering timeout by checking the worst case scenario from all of our tests. If I'm not mistaken, I believe that our worse case scenario, when it passes, is at about 50% of the current Engineering timeout.

Mocha's `before` and `beforeEach` are methods that allow you to specify functions that should be executed either before all of the tests run or before each test runs, respectively. Similarly, `after` and `afterEach` exist so that developers can specify functions that should be executed after all tests run or after each test runs, respectively.

We recommend using `beforeEach` rather than `before`, just as much as we recommend using `afterEach` rather than `after`. The idea is that each test case should not depend on a state that is shared with other tests. Use `beforeEach` to execute tasks that will prepare the resources needed for each test to run cleanly, and `afterEach` to tear down or clean those settings before the next test runs.

`before` and `after` can be used to define or create heavy resources that could cleanly be used by more than one test. We will see some examples below.

A discouraged example of `before` and `after` ([source](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/template/template/test/internal/beforeAfter.spec.ts#L12)):

```ts

> describe.skip("Discouraged example of `before` and `after`", function() {

Reviewer 4: skip ?

Author: This is discouraged! I don't want something discouraged to run 😁

Author: I've removed the skip here!
